misk-jobqueue-testing/src/main/kotlin/misk/jobqueue/FakeJobQueue.kt

```diff
@@ -150,14 +169,30 @@ class FakeJobQueue @Inject constructor(
       }
     }
 
+    // Reenqueue deadlettered jobs outside of the main loop to prevent an infinite loop.
     result.forEach { job ->
       if (job.deadLettered) {
-        deadletteredJobs.getOrPut(queueName, ::ConcurrentLinkedDeque).add(job)
+        deadletteredJobs.getOrPut(job.queueName, ::ConcurrentLinkedDeque).add(job)
       }
     }
-
     return result
   }
+
+  private fun pollNextJob(jobs: ConcurrentLinkedDeque<FakeJob>, considerDelays: Boolean): FakeJob? {
+    if (!considerDelays) {
+      return jobs.poll()
+    }
+    val now = clock.instant()!!
+    while (true) {
+      val job = jobs.firstOrNull {
+        it.deliveryDelay == null || !it.enqueuedAt.plus(it.deliveryDelay).isAfter(now)
+      }
+      // If removeFirstOccurrence() is false, then we lost race to another worker and should retry.
```

Comment:
Looks like `ConcurrentLinkedDeque` does not provide an atomic `find and remove`.